MerkleMountainRange/src/merklemountainrange.rs

```diff
@@ -0,0 +1,322 @@
+//      /  \   \ \      /  \   \
+//     /\  /\  /\ \    /\  /\  /\
+//    /\/\/\/\/\/\/\  /\/\/\/\/\/\
+//
+// Due to the unique way the MMR is constructed we can easily represent the MMR as a list of the nodes, as when
+// adding nodes you only append. Lets take the following MMR and number the nodes in the order we create them.
+//
+//        7
+//       /  \
+//      /    \
+//     3      6
+//    / \    / \
+//   1   2  4   5
+//
+// Looking above at the example of when you create the nodes, you will see the nodes will have been created in the
+// order as they are named. This means we can easily represent them as a list:
+// Height:  0 | 0 | 1 | 0 | 0 | 1 | 2
+// Node:    1 | 2 | 3 | 4 | 5 | 6 | 7
+//
+// Because of the list nature of the MMR we can easily navigate around the MMR using the following formulas:
+// Jump to peer : 2^(H+1) -1
+// find peak : 2^(H+1) -2 where < total elements
+// left down : 2^H
+// right down: -1
+// Note that the formulas are for direct indexes in the array, meaning the nodes count from 0 and not 1 as in
+// the examples above. H - Height
+// I - Index
+//
+// Pruning the MMR means removing flagging a node as pruned and only removing it if its peer has been removed.
+// We do this as we require the peer to prove the hash of the node. Taking the above example, let's prune leave 1.
+//
+//                 /\
+//                /  \
+//               /    \
+//              /      \
+//             /        \
+//            /          \
+//           /            \
+//          /\             \
+//         /\ \            /\
+//        /  \ \          /  \
+//       /\   \ \        /\   \
+//      /  \   \ \      /  \   \
+//     /\  /\  /\ \    /\  /\  /\
+//    /\/\/\/\/\/\/\  /\/\/\/\/\/\
+//
+// Node 1 has now only been marked as pruned but we cannot remove it as of yet because we still require it to
+// prove node 2. When we prune node 2, the MMR looks as follows
+//
+//                 /\
+//                /  \
+//               /    \
+//              /      \
+//             /        \
+//            /          \
+//           /            \
+//          /\             \
+//         /\ \            /\
+//        /  \ \          /  \
+//       /\   \ \        /\   \
+//      /  \   \ \      /  \   \
+//     /\  /\  /\ \    /\  /\  /\
+//      /\/\/\/\/\/\  /\/\/\/\/\/\
+//
+// Although we have not removed node 1 and node 2 from the MMR, we cannot yet remove node 3 as we require node 3 for
+// the proof of node 6. Let's prune 4 and 5.
+//
+//                 /\
+//                /  \
+//               /    \
+//              /      \
+//             /        \
+//            /          \
+//           /            \
+//          /\             \
+//         /\ \            /\
+//        /  \ \          /  \
+//       /\   \ \        /\   \
+//      /  \   \ \      /  \   \
+//         /\  /\ \    /\  /\  /\
+//        /\/\/\/\/\  /\/\/\/\/\/\
+//
+// Now we removed 3 from the MMR
```

Comment:
I don't think this background is entirely necessary here. It is very similar to the docs in the link, so the link should be sufficient by itself. Otherwise consider moving it to a README file